avocado/core/nrunner.py

```diff
@@ -680,6 +681,7 @@ class Task:
     def __init__(self, identifier, runnable, status_uris=None,
                  known_runners=None):
         self.identifier = identifier
+        self.uid = str(uuid4())
```

Comment:
The current implementation of the 'Task' class allows the command-line to set the identification of the task using the `-i` argument. See:

https://github.com/avocado-framework/avocado/blob/044b224c2ba3070e95e0805e130e2632b77a7968/avocado/core/nrunner.py#L717

```bash
avocado-runner task-run
usage: avocado-runner task-run [-h] -i IDENTIFIER [-s STATUS_URI] -k KIND [-u URI] [-c CONFIG] [-a ARG] [KEY_VAL ...]
avocado-runner task-run: error: the following arguments are required: -i/--identifier, -k/--kind
```

But the `identification` attribute is changed during test execution, making it difficult to query the status of the tasks. My idea with the `uuid` is to make it the default identifier of a task and change the current `identifier` attribute to `name`.

For consistency, I think it is worth changing the current command line argument from `-i/--identifier` to `-n/--name`. I understand this is a major change (command-line argument, API, self-tests), and, in this case, I prefer to make it on a different PR.

Let me know if that makes sense.

Reply:
> The current implementation of the 'Task' class allows the command-line to set the identification of the task using the `-i` argument. See:
> 
> https://github.com/avocado-framework/avocado/blob/044b224c2ba3070e95e0805e130e2632b77a7968/avocado/core/nrunner.py#L717
> 
> ```shell
> avocado-runner task-run
> usage: avocado-runner task-run [-h] -i IDENTIFIER [-s STATUS_URI] -k KIND [-u URI] [-c CONFIG] [-a ARG] [KEY_VAL ...]
> avocado-runner task-run: error: the following arguments are required: -i/--identifier, -k/--kind
> ```
> 
> But the `identification` attribute is changed during test execution, making it difficult to query the status of the tasks. My idea with the `uuid` is to make it the default identifier of a task and change the current `identifier` attribute to `name`.
> 

Just to be extra clear, you mean the `identifier` attribute, right?

Now, where is it that it gets changed? I see on `avocado/plugins/runner_nrunner.py`:

```python
    def _get_all_runtime_tasks(self, test_suite):
        result = []
        no_digits = len(str(len(test_suite)))
        for index, task in enumerate(test_suite.tests, start=1):
            task.known_runners = nrunner.RUNNERS_REGISTRY_PYTHON_CLASS
            # this is all rubbish data
            if test_suite.name:
                prefix = "{}-{}".format(test_suite.name, index)
            else:
                prefix = index
            test_id = TestID(prefix,
                             task.runnable.uri,
                             None,
                             no_digits)
            task.identifier = test_id
            result.append(RuntimeTask(task))
        return result
```

The command line argument `-i`, if my memory serves me well, is just given the identifier set on the code block copied above.

Let me know if I'm missing something.

Thanks! 

> For consistency, I think it is worth changing the current command line argument from `-i/--identifier` to `-n/--name`. I understand this is a major change (command-line argument, API, self-tests), and, in this case, I prefer to make it on a different PR.
> 
> Let me know if that makes sense.

Reply:
> > The current implementation of the 'Task' class allows the command-line to set the identification of the task using the `-i` argument. See:
> > https://github.com/avocado-framework/avocado/blob/044b224c2ba3070e95e0805e130e2632b77a7968/avocado/core/nrunner.py#L717
> > 
> > ```shell
> > avocado-runner task-run
> > usage: avocado-runner task-run [-h] -i IDENTIFIER [-s STATUS_URI] -k KIND [-u URI] [-c CONFIG] [-a ARG] [KEY_VAL ...]
> > avocado-runner task-run: error: the following arguments are required: -i/--identifier, -k/--kind
> > ```
> > 
> > 
> > But the `identification` attribute is changed during test execution, making it difficult to query the status of the tasks. My idea with the `uuid` is to make it the default identifier of a task and change the current `identifier` attribute to `name`.
> 
> Just to be extra clear, you mean the `identifier` attribute, right?
> 

Yes, I meant `identifier` and not `identification`, that was a typo.

> Now, where is it that it gets changed? I see on `avocado/plugins/runner_nrunner.py`:
> 
> ```python
>     def _get_all_runtime_tasks(self, test_suite):
>         result = []
>         no_digits = len(str(len(test_suite)))
>         for index, task in enumerate(test_suite.tests, start=1):
>             task.known_runners = nrunner.RUNNERS_REGISTRY_PYTHON_CLASS
>             # this is all rubbish data
>             if test_suite.name:
>                 prefix = "{}-{}".format(test_suite.name, index)
>             else:
>                 prefix = index
>             test_id = TestID(prefix,
>                              task.runnable.uri,
>                              None,
>                              no_digits)
>             task.identifier = test_id
>             result.append(RuntimeTask(task))
>         return result
> ```
> 

Yep, this is the exact part that the task identifier gets changed. I added some prints to show the result of this change:

```
[wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner passtest.py 
930c876c-85c7-11eb-9328-181dea939d92
requirement-package-vim-common
requirement-package-bla
JOB ID     : eaa6618223ffcd96f09af5438338428cae17e802
JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/job.log
 (2/3) avocado-software-manager: STARTED
 (3/3) avocado-software-manager: STARTED
 (2/3) avocado-software-manager: PASS (8.26 s)
 (3/3) avocado-software-manager: FAIL (9.38 s)
 (1/3) passtest.py:PassTest.test: STARTED
 (1/3) passtest.py:PassTest.test: PASS (0.01 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/results.html
JOB TIME   : 11.60 s
[wrampazz@wrampazz requirements_resolver]$
```

So, the requirements tasks are created with identifiers `requirement-package-vim-common` and `requirement-package-bla`, but after the tasks are already living in the state machine, their identifiers are changed to `avocado-software-manager` for both tasks, making it impossible to keep track of them.

Here is the `test-results` directory:

```
[wrampazz@wrampazz avocado.dev]$ ls /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/test-results/
1-passtest.py_PassTest.test  2-avocado-software-manager  3-avocado-software-manager
```

Although it has "different" identifiers, with the index as the prefix, if you look at the output of the job you will see it prints just `avocado-software-manager` for both tasks related to requirements.

I handle this problem using the new `uid` attribute as it is the same since the creation of the task.

To have better control of tasks, I think we should decouple the task id from the test id. Task ids should be related to the ids used inside the state machine only. I keep my suggestion here to have a uid for the tasks and have the state machine use it for controlling the tasks.

We can keep the other attribute as `identifier` but I think it is confusing to have `uid` and `identifier`, that is why I mentioned changing it to `name` in the previous version of this PR, but for consistency, it would be good to also change the `-i/--identifier` command-line parameter to `-n/--name`.

> The command line argument `-i`, if my memory serves me well, is just given the identifier set on the code block copied above.
> 

It is the same as the above example. If I use the command line, it works and set the `identifier` as expected:

```
[wrampazz@wrampazz requirements_resolver]$ avocado-runner-requirement task-run -i test -k requirement -u avocado-software-manager -a list
test
{'status': 'started', 'time': 33099.809073389, 'result': 'fail', 'output_dir': '/tmp/.avocado-task-ym00j4s_', 'id': 'test'}
{'status': 'running', 'time': 33099.829414541, 'result': 'fail', 'id': 'test'}
{'status': 'finished', 'time': 33100.067458345, 'returncode': 0, 'stdout': b'', 'stderr': b'', 'result': 'pass', 'id': 'test'}
[wrampazz@wrampazz requirements_resolver]$
```

But this `identifier` is changed by the code you mentioned.

Let me know if it answers your questions and if the proposal still makes sense.

> Let me know if I'm missing something.
> 
> Thanks!
> 
> > For consistency, I think it is worth changing the current command line argument from `-i/--identifier` to `-n/--name`. I understand this is a major change (command-line argument, API, self-tests), and, in this case, I prefer to make it on a different PR.
> > Let me know if that makes sense.

Reply:
> Yep, this is the exact part that the task identifier gets changed. I added some prints to show the result of this change:
> 
> ```
> [wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner passtest.py 
> 930c876c-85c7-11eb-9328-181dea939d92
> requirement-package-vim-common
> requirement-package-bla
> JOB ID     : eaa6618223ffcd96f09af5438338428cae17e802
> JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/job.log
>  (2/3) avocado-software-manager: STARTED
>  (3/3) avocado-software-manager: STARTED
>  (2/3) avocado-software-manager: PASS (8.26 s)
>  (3/3) avocado-software-manager: FAIL (9.38 s)
>  (1/3) passtest.py:PassTest.test: STARTED
>  (1/3) passtest.py:PassTest.test: PASS (0.01 s)
> RESULTS    : PASS 2 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
> JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/results.html
> JOB TIME   : 11.60 s
> [wrampazz@wrampazz requirements_resolver]$
> ```
> 
> So, the requirements tasks are created with identifiers `requirement-package-vim-common` and `requirement-package-bla`, but after the tasks are already living in the state machine, their identifiers are changed to `avocado-software-manager` for both tasks, making it impossible to keep track of them.

OK, so the problem is really how we're keeping track of them.  Even though the task's `identifier` attributes are being changed, the tasks themselves are the same.  I see no reason for not using the *tasks* as prereqs, instead of task's identifiers.  This tracking is going to be done all in the same job/state machine, all running on the same Python interpreter.

This commit is a working version of what I described above: https://github.com/clebergnu/avocado/commit/1439bd7a3f084a199fc42dac073133f6268f4c35 

> 
> Here is the `test-results` directory:
> 
> ```
> [wrampazz@wrampazz avocado.dev]$ ls /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/test-results/
> 1-passtest.py_PassTest.test  2-avocado-software-manager  3-avocado-software-manager
> ```
> 

And now given the fact that you brought the test results to the discussion, what are your plans with regards to the handling of tasks that really represent tests, and tasks that are requirements?

> Although it has "different" identifiers, with the index as the prefix, if you look at the output of the job you will see it prints just `avocado-software-manager` for both tasks related to requirements.
> 
> I handle this problem using the new `uid` attribute as it is the same since the creation of the task.
> 
> To have better control of tasks, I think we should decouple the task id from the test id. Task ids should be related to the ids used inside the state machine only. I keep my suggestion here to have a uid for the tasks and have the state machine use it for controlling the tasks.
> 
> We can keep the other attribute as `identifier` but I think it is confusing to have `uid` and `identifier`, that is why I mentioned changing it to `name` in the previous version of this PR, but for consistency, it would be good to also change the `-i/--identifier` command-line parameter to `-n/--name`.
> 
> > The command line argument `-i`, if my memory serves me well, is just given the identifier set on the code block copied above.
> 
> It is the same as the above example. If I use the command line, it works and set the `identifier` as expected:
> 
> ```
> [wrampazz@wrampazz requirements_resolver]$ avocado-runner-requirement task-run -i test -k requirement -u avocado-software-manager -a list
> test
> {'status': 'started', 'time': 33099.809073389, 'result': 'fail', 'output_dir': '/tmp/.avocado-task-ym00j4s_', 'id': 'test'}
> {'status': 'running', 'time': 33099.829414541, 'result': 'fail', 'id': 'test'}
> {'status': 'finished', 'time': 33100.067458345, 'returncode': 0, 'stdout': b'', 'stderr': b'', 'result': 'pass', 'id': 'test'}
> [wrampazz@wrampazz requirements_resolver]$
> ```
> 
> But this `identifier` is changed by the code you mentioned.
> 
> Let me know if it answers your questions and if the proposal still makes sense.
> 

Please take a look at the this branch: https://github.com/clebergnu/avocado/commits/requirements_resolver_with_task_identifier

And let me know if you see any issues with it.

Reply:
> > Yep, this is the exact part that the task identifier gets changed. I added some prints to show the result of this change:
> > ```
> > [wrampazz@wrampazz requirements_resolver]$ avocado run --test-runner=nrunner passtest.py 
> > 930c876c-85c7-11eb-9328-181dea939d92
> > requirement-package-vim-common
> > requirement-package-bla
> > JOB ID     : eaa6618223ffcd96f09af5438338428cae17e802
> > JOB LOG    : /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/job.log
> >  (2/3) avocado-software-manager: STARTED
> >  (3/3) avocado-software-manager: STARTED
> >  (2/3) avocado-software-manager: PASS (8.26 s)
> >  (3/3) avocado-software-manager: FAIL (9.38 s)
> >  (1/3) passtest.py:PassTest.test: STARTED
> >  (1/3) passtest.py:PassTest.test: PASS (0.01 s)
> > RESULTS    : PASS 2 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
> > JOB HTML   : /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/results.html
> > JOB TIME   : 11.60 s
> > [wrampazz@wrampazz requirements_resolver]$
> > ```
> > 
> > 
> > So, the requirements tasks are created with identifiers `requirement-package-vim-common` and `requirement-package-bla`, but after the tasks are already living in the state machine, their identifiers are changed to `avocado-software-manager` for both tasks, making it impossible to keep track of them.
> 
> OK, so the problem is really how we're keeping track of them. Even though the task's `identifier` attributes are being changed, the tasks themselves are the same. I see no reason for not using the _tasks_ as prereqs, instead of task's identifiers. This tracking is going to be done all in the same job/state machine, all running on the same Python interpreter.
> 
> This commit is a working version of what I described above: [clebergnu@1439bd7](https://github.com/clebergnu/avocado/commit/1439bd7a3f084a199fc42dac073133f6268f4c35)
> 

This looks interesting and I don't see a reason for not using the task object for tracking. I'll try it with my tests.

One thing that came to my mind is if keeping the `identifier` attribute with this name won't confuse new developers for the code maintenance in the future.

> > Here is the `test-results` directory:
> > ```
> > [wrampazz@wrampazz avocado.dev]$ ls /home/wrampazz/avocado/job-results/job-2021-03-15T16.49-eaa6618/test-results/
> > 1-passtest.py_PassTest.test  2-avocado-software-manager  3-avocado-software-manager
> > ```
> 
> And now given the fact that you brought the test results to the discussion, what are your plans with regards to the handling of tasks that really represent tests, and tasks that are requirements?
> 

Speaking about the `test-results` dir, being able to keep the `identifier` attribute unchanged would help here. This was going to be my proposal for the future refactor and extension of the `identifier` attribute.

I imagine that in a near future, we will have more and more actions being handled as tasks, like, for example, preparing a container image, checking the availability of a virtual machine, or a remote host.

I imagine the task will be the `core` of things running on nrunner and, in this case, we will need a good management of tasks to make logging and debugging easy.

> > Although it has "different" identifiers, with the index as the prefix, if you look at the output of the job you will see it prints just `avocado-software-manager` for both tasks related to requirements.
> > I handle this problem using the new `uid` attribute as it is the same since the creation of the task.
> > To have better control of tasks, I think we should decouple the task id from the test id. Task ids should be related to the ids used inside the state machine only. I keep my suggestion here to have a uid for the tasks and have the state machine use it for controlling the tasks.
> > We can keep the other attribute as `identifier` but I think it is confusing to have `uid` and `identifier`, that is why I mentioned changing it to `name` in the previous version of this PR, but for consistency, it would be good to also change the `-i/--identifier` command-line parameter to `-n/--name`.
> > > The command line argument `-i`, if my memory serves me well, is just given the identifier set on the code block copied above.
> > 
> > 
> > It is the same as the above example. If I use the command line, it works and set the `identifier` as expected:
> > ```
> > [wrampazz@wrampazz requirements_resolver]$ avocado-runner-requirement task-run -i test -k requirement -u avocado-software-manager -a list
> > test
> > {'status': 'started', 'time': 33099.809073389, 'result': 'fail', 'output_dir': '/tmp/.avocado-task-ym00j4s_', 'id': 'test'}
> > {'status': 'running', 'time': 33099.829414541, 'result': 'fail', 'id': 'test'}
> > {'status': 'finished', 'time': 33100.067458345, 'returncode': 0, 'stdout': b'', 'stderr': b'', 'result': 'pass', 'id': 'test'}
> > [wrampazz@wrampazz requirements_resolver]$
> > ```
> > 
> > 
> > But this `identifier` is changed by the code you mentioned.
> > Let me know if it answers your questions and if the proposal still makes sense.
> 
> Please take a look at the this branch: https://github.com/clebergnu/avocado/commits/requirements_resolver_with_task_identifier
> 
> And let me know if you see any issues with it.

Sure, I'll take a look, run some tests and let you know.

Reply:
> Please take a look at the this branch: https://github.com/clebergnu/avocado/commits/requirements_resolver_with_task_identifier
> 
> And let me know if you see any issues with it.

I tested and evaluated the branch. From a test point of view, it worked without problems. From the code point of view, my only concern is this commit: https://github.com/clebergnu/avocado/commit/10499c113da9469aa6cc414cb34fc8801ec0b488

With it, all instantiations of a `Task` in the nrunner code are passing the `identifier` as `None`, making it useless as an argument. If we cannot set it, as it will be changed later in the code, I don't see a reason to let it as an argument.

The Task `identifier` can be a valuable attribute for logging in a human-readable way. As you mentioned before, using the object for the Task identification inside the state machine is fine and works without problems, but also having a human-readable attribute for logging makes things easy.

For the next PR version, I'll incorporate the other changes you suggested related to the object identification inside the state machine and I'll try to propose a way to keep the task `identifier` unchanged if we are handling anything other than a test. What do you think?

Reply:
@clebergnu I'm trying to understand why not adopt the @willianrampazzo suggestion here.

Having `uuid` and `identifier` it looks redundant and confusing, also, having only an "identifier" that is getting changed during execution seems strange to me, too. So having a UUID and a name seems logical here, IMO.

I like the idea of having references to objects, but I don't think this invalidates the @willianrampazzo's proposal, for better organization and readability.

Could you ELI5 please? Maybe I'm missing something ;)

Reply:
> @clebergnu I'm trying to understand why not adopt the @willianrampazzo suggestion here.
> 
> Having `uuid` and `identifier` it looks redundant and confusing, also, having only an "identifier" that is getting changed during execution seems strange to me, too. So having a UUID and a name seems logical here, IMO.
> 
> I like the idea of having references to objects, but I don't think this invalidates the @willianrampazzo's proposal, for better organization and readability.
> 
> Could you ELI5 please? Maybe I'm missing something ;)

The `Task`'s identifier is needed for the execution of a task, and that, in the context of an Avocado job, is needed very late.   My initial instinctive feeling was that changing the identifier could be a corner case, and maybe could be avoided.

So I went through the code and "questioned" the status quo, as in, "why are we creating the Task so earlier, and then changing its identifier?".  It was clear that there was no good reason, besides code that was as old as the old `nrun` plugin (which did not implement integration with an Avocado job).

So I went through the `Runnable` and `Task` responsibilities. Not only the `Task`'s identifier, but the `Task` itself, contains *execution time* information.  Fortunately there's a good separation of responsibilities, and the Runnable class describes *what* needs to be run.  This is all we need in a suite.

So the result of that investigation is in PR #4469.  Please let me know if that makes sense to you.

Reply:
> > Please take a look at the this branch: https://github.com/clebergnu/avocado/commits/requirements_resolver_with_task_identifier
> > And let me know if you see any issues with it.
> 
> I tested and evaluated the branch. From a test point of view, it worked without problems. From the code point of view, my only concern is this commit: [clebergnu@10499c1](https://github.com/clebergnu/avocado/commit/10499c113da9469aa6cc414cb34fc8801ec0b488)
> 
> With it, all instantiations of a `Task` in the nrunner code are passing the `identifier` as `None`, making it useless as an argument. If we cannot set it, as it will be changed later in the code, I don't see a reason to let it as an argument.
> 
> The Task `identifier` can be a valuable attribute for logging in a human-readable way. As you mentioned before, using the object for the Task identification inside the state machine is fine and works without problems, but also having a human-readable attribute for logging makes things easy.
> 
> For the next PR version, I'll incorporate the other changes you suggested related to the object identification inside the state machine and I'll try to propose a way to keep the task `identifier` unchanged if we are handling anything other than a test. What do you think?

@willianrampazzo please let me know if the explanation given earlier to @beraldoleal makes sense (and specially PR #4469 ).  Thanks!